src/main/java/api/json/GoblintSocketMessageProducer.java

```diff
@@ -0,0 +1,123 @@
+ */
+
+public class GoblintSocketMessageProducer implements MessageProducer, Closeable, MessageConstants {
+
+    private static final Logger LOG = Logger.getLogger(StreamMessageProducer.class.getName());
+
+    private final MessageJsonHandler jsonHandler;
+    private final MessageIssueHandler issueHandler;
+    private final InputStream input;
+
+    private MessageConsumer callback;
+    private boolean keepRunning;
+
+    private final org.apache.logging.log4j.Logger log = LogManager.getLogger(GoblintSocketMessageConsumer.class);
+
+    public GoblintSocketMessageProducer(InputStream input, MessageJsonHandler jsonHandler) {
+        this(input, jsonHandler, null);
+    }
+
+    public GoblintSocketMessageProducer(InputStream input, MessageJsonHandler jsonHandler, MessageIssueHandler issueHandler) {
+        this.input = input;
+        this.jsonHandler = jsonHandler;
+        this.issueHandler = issueHandler;
+    }
+
+    @Override
+    public void listen(MessageConsumer callback) {
+        if (keepRunning) {
+            throw new IllegalStateException("This StreamMessageProducer is already running.");
+        }
+        this.keepRunning = true;
+        this.callback = callback;
+        try {
+            while (keepRunning) {
+                boolean result = handleMessage(input);
+                if (!result)
+                    keepRunning = false;
+            } // while (keepRunning)
+        } catch (IOException exception) {
+            if (JsonRpcException.indicatesStreamClosed(exception)) {
+                // Only log the error if we had intended to keep running
+                if (keepRunning)
+                    fireStreamClosed(exception);
+            } else
+                throw new JsonRpcException(exception);
+        } finally {
+            this.callback = null;
+            this.keepRunning = false;
+        }
+    }
+
+    /**
+     * Log an error.
+     */
+    protected void fireError(Throwable error) {
+        String message = error.getMessage() != null ? error.getMessage() : "An error occurred while processing an incoming message.";
+        LOG.log(Level.SEVERE, message, error);
+    }
+
+    /**
+     * Report that the stream was closed through an exception.
+     */
+    protected void fireStreamClosed(Exception cause) {
+        String message = cause.getMessage() != null ? cause.getMessage() : "The input stream was closed.";
+        LOG.log(Level.INFO, message, cause);
+    }
+
+
+    /**
+     * Read the JSON content part of a message, parse it, and notify the callback.
+     *
+     * @return {@code true} if we should continue reading from the input stream, {@code false} if we should stop
+     */
+    protected boolean handleMessage(InputStream input) throws IOException {
+        if (callback == null)
+            callback = message -> LOG.log(Level.INFO, "Received message: " + message);
+
+        try {
+            BufferedReader inputReader = new BufferedReader(new InputStreamReader(input));

```

Comment:
Only one `BufferedReader` should be created (e.g. in constructor) instead of recreating one every time. Otherwise data could be lost.